pgrx-pg-sys/src/submodules/elog.rs

```diff
@@ -103,14 +103,109 @@ impl From<i32> for PgLogLevel {
 ///
 /// The output these logs goes to the PostgreSQL log file at `DEBUG5` level, depending on how the
 /// [PostgreSQL settings](https://www.postgresql.org/docs/current/runtime-config-logging.html) are configured.
+///
+/// ## Adding Details and Hints
+///
+/// Additionally, you can use specify a `detail` and/or `hint` to be included with the report.
+/// After the string literal to be used as the message and any formatting arguments, add a semicolon.
+/// Then assign a string literal to `detail` or `hint`, optionally followed by a list of
+/// comma separated formatting arguments.
+///
+/// ## Examples
+///
+/// ```rust,no_run
+/// use pgrx_pg_sys::debug5;
+///
+/// // just like `println!` and `format!`
+/// debug5!("a simple message");
+/// debug5!("or a formatted message: {:?}", "pgrx rocks!");
+/// // include details and hints
+/// debug5! {
+///     "add details if you want";
+///     detail = "...";
+/// }
+/// debug5! {
+///     "or a message with just a hint";
+///     hint = "...";
+/// }
+/// debug5! {
+///     "put it all together for {} {}", "great", "success!";
+///     detail = "extra {}", "info";
+///     hint = "{} helpful", "very";
+/// }
+/// ```
 #[macro_export]
 macro_rules! debug5 {
+    (
+        $msg:literal $(, $msg_args:expr)*;
+        $(detail = $detail:literal $(, $detail_args:expr)*)?;
+    ) => (
+        {
+            extern crate core;
+            extern crate alloc;
+            $crate::ereport!(
+                $crate::elog::PgLogLevel::DEBUG5,
+                $crate::errcodes::PgSqlErrorCode::ERRCODE_SUCCESSFUL_COMPLETION,
+                alloc::format!("{}", core::format_args!($msg $(, $msg_args)*));
+                $(detail = alloc::format!("{}", core::format_args!($detail $(, $detail_args)*));)?
+            );
+        }
+    );
+    (
+        $msg:literal $(, $msg_args:expr)*;
+        $(hint = $hint:literal $(, $hint_args:expr)*)?;
+    ) => (
+        {
+            extern crate core;
+            extern crate alloc;
+            $crate::ereport!(
+                $crate::elog::PgLogLevel::DEBUG5,
+                $crate::errcodes::PgSqlErrorCode::ERRCODE_SUCCESSFUL_COMPLETION,
+                alloc::format!("{}", core::format_args!($msg $(, $msg_args)*));
+                $(hint = alloc::format!("{}", core::format_args!($hint $(, $hint_args)*));)?
+            );
+        }
+    );
+    (
+        $msg:literal $(, $msg_args:expr)*;
+        $(hint = $hint:literal $(, $hint_args:expr)*)?;
+        $(detail = $detail:literal $(, $detail_args:expr)*)?;
+    ) => (
```

Comment:
This is a somewhat deceptive syntax. It looks simple, but it actually has a fair amount of nuance. For instance, this introduces a significant divergence between the behavior in the presence of a `;` versus not, right?

I'll sleep on it, but it definitely can't be merged without UI tests covering pathological inputs, like
```rust
debug5!(hint = "..."; detail = "...";);
debug5!(detail = "...");
```

You can find the existing ui tests in `./pgrx-tests/tests/ui`. They allow us to cover erroneous cases that don't compile. I would prefer that the patterns that worked with 0.11.2 continue to work in 0.12.0, and that anything that seems weird or ambiguous fails.

Also, this code has a rather lot of duplication now. That might have been tolerable when there was less code overall, but I should note that Rust allows you to define a macro that itself mostly serves to define new macros.

Reply:
After considering your feedback and reworking some things, I decided to change it to `message = $msg:literal`. I like this makes the syntax more obvious and helped to avoid some of the ambiguity when matching. I also added ui tests, please take a look and let me know if/how they can be improved.